packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -442,28 +451,32 @@ class OverlayState extends State<Overlay> with TickerProviderStateMixin {
 
   @override
   Widget build(BuildContext context) {
-    // These lists are filled backwards. For the offstage children that
-    // does not matter since they aren't rendered, but for the onstage
-    // children we reverse the list below before adding it to the tree.
-    final List<Widget> onstageChildren = <Widget>[];
-    final List<Widget> offstageChildren = <Widget>[];
+    // This list is filled backwards and then reversed below before
+    // it is added to the tree.
+    final List<Widget> children = <Widget>[];
     bool onstage = true;
+    int onstageCount = 0;
     for (int i = _entries.length - 1; i >= 0; i -= 1) {
       final OverlayEntry entry = _entries[i];
       if (onstage) {
-        onstageChildren.add(_OverlayEntry(entry));
+        onstageCount += 1;
+        children.add(_OverlayEntryWidget(
+          key: entry._key,
+          entry: entry,
+        ));
         if (entry.opaque)
           onstage = false;
       } else if (entry.maintainState) {
-        offstageChildren.add(TickerMode(enabled: false, child: _OverlayEntry(entry)));
+        children.add(_OverlayEntryWidget(
+          key: entry._key,
+          entry: entry,
+          tickerEnabled: false,
+        ));
       }
     }
     return _Theatre(
-      onstage: Stack(
-        fit: StackFit.expand,
-        children: onstageChildren.reversed.toList(growable: false),
-      ),
-      offstage: offstageChildren,
+      skipCount: children.length - onstageCount,
```

Comment:
This seems like it would fail if the logic above added children in an order like [offstage, onstage, offstage, offstage] for some reason.

If we think that's invalid, we should probably assert it. If it's actually valid we need different logic here.

Reply:
Doing something like [offstage, onstage, offstage, offstage] is not supported by the overlay. The top x entries are always onstage, everything else is either completely ignored or - if maintainState is true - rendered offstage.

Not sure what we should assert here. The assert would basically duplicate the algorithm above.